src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -131,33 +115,43 @@ where s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id)
 
         ConcurrentDictionary<Guid, VersionedSagaEntity> data = new ConcurrentDictionary<Guid, VersionedSagaEntity>();
 
-        int version;
-
         class VersionedSagaEntity
         {
             public VersionedSagaEntity(IContainSagaData sagaData, VersionedSagaEntity original = null)
             {
                 SagaData = DeepClone(sagaData);
-                versionCache = original?.versionCache ?? new ConditionalWeakTable<IContainSagaData, SagaVersion>();
+                if (original != null)
+                {
+                    original.ConcurrencyCheck(sagaData);
+
+                    versionCache = original.versionCache;
+                    version = original.version;
+                    Interlocked.Increment(ref version);
```

Comment:
Why is this `Interlocked.Increment` instead of `version++` ? Nothing can access newly created `VersionedSagaEntity` before `AddOrUpdate` finishes hence why `Interlocked.Increment` is used?